### PR TITLE
Stereoscope lit

### PR DESCRIPTION
**Pull request by romain-lopez**
These changes make it possible to run Stereoscope with the new API. It will require some improvements before merging. So far, Adam and I have flagged:
+ documentation (overall)
+ math in documentation can be formated
+ using register buffer instead of parameter with no grad
+ add reference in documentation
+ save / load in unit test? 

Please let me know if you find some other things.

### CODE REVIEW

**Comment by romain-lopez** [comment]
Ah! and we might want to look into this trainer behavior for no validation set and see if we can do something nicer. I don't like having to give the train_set twice to the pytorch lightning task

File: scvi/external/stereoscope/_module.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    @torch.no_grad()
+    def sample(
+        self,
+        tensors,
+        n_samples=1,
+        library_size=1,
+    ):
+    # not a generative model
+        return None
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:239]: raise NotImplementedError

File: scvi/model/base/_base_model.py [hunk 2]
```diff
@@ -272,7 +272,11 @@ def train(
         task_kwargs = vae_task_kwargs if isinstance(vae_task_kwargs, dict) else dict()
         self._pl_task = task_class(self.model, len(self.train_indices_), **task_kwargs)
 
-        self.trainer.fit(self._pl_task, train_dl, val_dl)
+        if train_size==1.0:
+            # circumvent the empty DL problem
+            self.trainer.fit(self._pl_task, train_dl, train_dl)
```
Comment (adamgayoso) [inline scvi/model/base/_base_model.py:277]: I think what you can do is just pass the train_dl and that's it, no val loop this way.

File: tests/external/test_stereoscope.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import scvi
```
Comment (adamgayoso) [inline tests/external/test_stereoscope.py:1]: this needs to be a proper test, see the examples in test_models.py, just needs to be in function called `test_stereoscope()`.

You should also use the synthetic iid dataset for simplicity.

File: tests/external/test_stereoscope.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+            save_path="tests/data/10X",
+            remove_extracted_data=True,
+            run_setup_anndata=True,
+        )
+
+
+model = RNAStereoscope(dataset)
+model.train(n_epochs=100, frequency=1, vae_task_kwargs={"lr":0.01}, train_size=1.0)
+params = model.get_params()
+print(model.history)
```
Comment (adamgayoso) [inline tests/external/test_stereoscope.py:18]: no need for print, could just smoke test it.

File: tests/external/test_stereoscope.py [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import scvi
+import os
+
+import numpy as np
+from scvi.external.stereoscope import RNAStereoscope, SpatialStereoscope
```
Comment (adamgayoso) [inline tests/external/test_stereoscope.py:5]: you shouldn't need `scvi.external.stereoscope` -- it should be just `scvi.external`.

File: scvi/external/stereoscope/_model.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+    Reimplementation of Stereoscope for deconvolution of spatial transcriptomics from single-cell transcriptomics.
+    
+    https://github.com/almaan/stereoscope.
+    Parameters
+    ----------
+    sc_adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    use_cuda
+        Use the GPU or not.
+
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:22]: need docstring for model kwargs -- see the scvi.py code for how to link to the proper model class.

File: scvi/external/stereoscope/_model.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:76]: need model kwargs here too

File: scvi/external/stereoscope/_model.py [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+    Examples
+    --------
+    >>> st_adata = anndata.read_h5ad(path_to_st_anndata)
+    >>> scvi.data.setup_anndata(st_adata)
+    >>> st_adata.obs["indices"] = np.arange(st_adata.n_obs)
+    >>> register_tensor_from_anndata(st_adata, "ind_x", "obs", "indices")
+    >>> stereo = scvi.external.stereoscope.SpatialStereoscope(st_adata, sc_params)
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:83]: example shouldn't need stereoscope -- should be `scvi.external.SpatialStereoscope`, same for RNA

File: scvi/external/stereoscope/_module.py [hunk 9]
```diff
@@ -0,0 +1,240 @@
+from scvi.compose import AbstractVAE, auto_move_data, SCVILoss
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:1]: the imports should be sorted. https://pypi.org/project/isort/

just be sure to isort only the stereoscope directory

File: scvi/external/stereoscope/_module.py [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        prior_weight:Literal["n_obs", "minibatch"] = "n_obs",
+    ):
+        super().__init__()
+        self.W = torch.nn.Parameter(torch.tensor(params[0]), requires_grad=False)
+        self.px_o = torch.nn.Parameter(torch.tensor(params[1]), requires_grad=False)
+        self.n_spots = n_spots
+        self.n_genes, self.n_labels = self.W.shape
+        self.prior_weight = prior_weight
+
+        #####
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:130]: this math could actually go in the class docstring, if you look at the numpydoc specification, there's a notes section and you can do block math latex.

Comment (adamgayoso) [reply scvi/external/stereoscope/_module.py:130]: this applies to all the math if you wish.

File: scvi/external/stereoscope/_model.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+    sc_adata
+        single-cell AnnData object that has been registered via :func:`~scvi.data.setup_anndata`.
+    use_cuda
+        Use the GPU or not.
+
+    Examples
+    --------
+    >>> sc_adata = anndata.read_h5ad(path_to_sc_anndata)
+    >>> scvi.data.setup_anndata(sc_adata, label_key="labels")
+    >>> stereo = scvi.external.stereoscope.RNAStereoscope(sc_adata)
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:27]: no `stereoscope` needed here.

File: scvi/external/stereoscope/_module.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
...
+    @auto_move_data
+    def inference(self):
+        return {}
+
+    @auto_move_data
+    def generative(
+        self, x, ind_x
+    ):
+        """Build the deconvolution model for every cell in the minibatch."""
+        # x_sg \sim NB(softplus(\beta_g) * \sum_{z=1}^Z softplus(v_sz) * softplus(W)_gz + \gamma_s \eta_g, exp(px_r))
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:184]: math could be moved to docstring, see other comment

**Review by adamgayoso — CHANGES_REQUESTED** [review]

**Comment by romain-lopez** [comment]
Alright! I think this is good to go, and should have the minimal functionalities we discussed! Let me know if you have any other comments, and generally, if you see some obvious doc changes you'd like to see implemented please go ahead!

**Comment by romain-lopez** [comment]
checked pytest, flake8, black. not sure what codacy wants from me :)

File: tests/external/test_stereoscope.py [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import numpy as np
+
+from scvi.data import register_tensor_from_anndata, synthetic_iid
+from scvi.external import RNAStereoscope, SpatialStereoscope
+
+
+def test_scvi(save_path):
```
Comment (adamgayoso) [inline tests/external/test_stereoscope.py:7]:
```suggestion
def test_stereoscope(save_path):
```

**Review by adamgayoso — CHANGES_REQUESTED** [review]

File: scvi/external/stereoscope/_model.py [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+        )
+        self._model_summary_string = (
+            "RNADeconv Model with params: \nn_genes: {}, n_labels: {}"
+        ).format(
+            self.n_genes,
+            self.n_labels,
+        )
+        self.init_params_ = self._get_init_params(locals())
+
+    def get_params(self):
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:54]: need some sort of documentation here and typing

File: scvi/external/stereoscope/_model.py [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    >>> register_tensor_from_anndata(st_adata, "ind_x", "obs", "indices")
+    >>> stereo = scvi.external.SpatialStereoscope(st_adata, sc_params)
+    >>> stereo.train()
+    >>> st_adata.obs["deconv"] = stereo.get_proportions()
+    """
+
+    def __init__(
+        self,
+        st_adata: AnnData,
+        params: np.ndarray,
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_model.py:92]: need to add this to the docstring and add typing.

File: scvi/external/stereoscope/_module.py [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+        self.n_genes = n_genes
+        self.n_labels = n_labels
+
+        # logit param for negative binomial
+        self.px_o = torch.nn.Parameter(torch.randn(self.n_genes))
+        self.W = torch.nn.Parameter(
+            torch.randn(self.n_genes, self.n_labels)
+        )  # n_genes, n_cell types
+
+    def get_params(self) -> torch.Tensor:
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:40]: I'd add an `@torch.no_grad()` decorator here and then you don't need detach.

File: scvi/external/stereoscope/_module.py [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+        self.prior_weight = prior_weight
+
+        # noise from data
+        self.eta = torch.nn.Parameter(torch.randn(self.n_genes))
+        # factor loadings
+        self.V = torch.nn.Parameter(torch.randn(self.n_labels + 1, self.n_spots))
+        # additive gene bias
+        self.beta = torch.nn.Parameter(0.1 * torch.randn(self.n_genes))
+
+    def get_proportions(self, keep_noise=False) -> torch.Tensor:
```
Comment (adamgayoso) [inline scvi/external/stereoscope/_module.py:142]: same with torch no grad decorator

**Review by adamgayoso — CHANGES_REQUESTED** [review]

**Comment by romain-lopez** [comment]
added those!

**Comment by romain-lopez** [comment]
Actually, one more thing is how we should incorportate the registration of the ind_x into the anndata at load time with mininal changes. Any suggestions?
